# RNAseqEval
A collection of tools for evaluating RNA seq mapping. The tools were used to benchmark some RNA mapping tools on real and synthetic third generation sequencing datasets. Detailed description of synthetic dataset preparation can be found at [RNAseq_benchmark/data_preparation.md](RNAseq_benchmark/data_preparation.md). Description of the evaluation process for real and simulated datasets can be found at [RNAseq_benchmark/RNAseq_benchmark.md](RNAseq_benchmark/RNAseq_benchmark.md).

## generate_transcriptome.py
Run generate_transcriptome.py to generate a trancriptome from a reference and a set of annotations.

How to run:
  
     generate_transcriptome.py annotations.gtf reference.fasta transcriptome.fasta

Generated transcriptome can be used to simulate RNA reads using DNA simulators such as PBSIM.

## RNAseqEval.py
Run RNAseqEval.py for general evaulation of mappings in sam file against reference and optionally annotations. This script is intended to evaluate real dataset mapping. Run RNAseqEval.py without any arguments to print options.

Usage:
     
    RNAseqEval.py eval-mapping <reference FASTA file> <input SAM file> options

    options:
    -a <file> : a reference annotation (GFF/GTF/BED) file
    -o (--output) <file> : output file to which the report will be written

Detailed description of running modes, options and output can be found at [doc/RNAseqEval.md](doc/RNAseqEval.md).

## Process_pbsim_data.py
Run Process_pbsim_data.py to evaluate mappings of data generated by the PBSIM simulator. It requires the simulation data to be organized in a specific way.

How to run:
 
    Process_pbsim_data.py process simulation_root_folder mappings.sam annotations.gtf

Detailed description of options, output and required simulation data organization can be found at [doc/Process_pbsim_data.md](doc/Process_pbsim_data.md).
