# Process_pbsim_data.py
Run Process_pbsim_data.py to evaluate mappings of data generated by PBSIM simulator.

Usage example:
 
    Process_pbsim_data.py process simulation_root_folder mappings.sam annotations.bed

## Evaluation method
Process_pbsim_data.py evaluates the quality of mapping by comparing it against the simulation daa generated by PBSIM. Aside from simulation parameters, PBSIM will take as input a fasta file containing one or more references. PBSIM will use all references to generate simulated reads. For each reference it will generate 3 files:
1. REF file containing the reference
2. FASTQ file containing simulated reads
3. MAF file containing the information from which part of the reference was each read generated



## Evaluating a complex simulated dataset
The script is written so that it can evaluate a file obtained from several simulations. Each simulation can use different coverage, but needs to use the same reference genome. Reads simulated within a single simulation need to have the same FASTA header prefix. All PBSIM output folders (for different simulations) must be within the simulation_root_folder, and a dictionary simFolderDict (inside Process_pbsim_data.py script) should be modified to connect each simualted read header with the corresponding pbsim output folder.

## Example
Lets assume that we use three simulations, Pbsim output folders are named group1, group2 and group3, and header prefixes for eache simulation's reads are SimG1, SimG2 and SimG3 respectively.
Then all three simulation folders must be placed directly within simulation_root_folder and the dictionary simFolderDict, within the Process_pbsim_data.py script, must be definied as:

    simFolderDict = {'SimG1' : 'group1'
                   , 'SimG2' : 'group2'
                   , 'SimG3' : 'group3'}

## Output
Process_pbsim_daty.py scripts generates a report containing following information:

    Number of original SAM lines (in the input SAM file)
    Number of usable whole alignments (with valid CIGAR string)
    Number of annotations
    Number of multiexon genes

    Number of reads aligned to a start of an exon
    Number of reads aligned to an end of an exon
    Number of reads aligned to both, start and end of an exon
    Number of good whole alignments

    Number of alignments that hit both ends
    Number of alignments that hit all parts
    Number of alignments that hit at least one part
    Number of alignments that equals at least one part

    Number of split reads
    Number of alignments that hit both ends, for SPLIT reads only
    Number of alignments that hit all parts, for SPLIT reads only
    Number of alignments that hit at least one part, SPLITfor SPLIT reads only
    Number of alignments that equals at least one part, for SPLIT reads only
