# Example dataset
This folder contains a small example dataset that can be used to test our evaluation tools. Sample dataset contains only one group, and was simulated using PBSIM on _D. Melanogaster_ chromosome 4, with coverage 5 and with PacBio ROI parameters. Simulation files (generated by PBSIM) are placed in a folder __group1__, inside a root simulation folder __simulation\_X5__. Headers in the generated fastq file were adjusted as described in [RNAseq_benchmark/Data_preparation.md](/RNAseq_benchmark/Data_preparation.md)

Sample dataset contains 2955 reads.

The folder also contains all files needed to test the dataset and the tools.

Contents of the folder:
- example.fastq - small test dataset, 
- example.sam   - mapping of the example dataset to the reference genome using GraphMap (https://github.com/isovic/graphmap) and it option to map RNA reads with the help of annotations.
- dmelanogaster_chr_genome.fa  - genomic sequence for _D. Melanogaster_ chromosome 4, this represents a reference genome for our example dataset
- dmelanogaster_chr.gft - gene annotations for _D. Melanogaster_ chromosome 4
- dmelanogaster_chr_trans_F.fa - transcriptome for for _D. Melanogaster_ chromosome 4, this is generated from geneomic sequence and annotations using [generate_transcriptome.py](/generate_transcriptome.py). Designation '\_F' means that the transcriptome has been filtered, removing sequences shorter than 100 base-pairs, as required by PBSIM.

### Determining match rate
Matchrate for SAM file contining mappings can be determined using errorrates.py script from https://github.com/isovic/samscripts.

    python /home/kkrizanovic/src/samscripts/src/errorrates.py base dmelanogaster_chr4_genome.fa sample.sam

__Results:__

                       	mean	std	    median	min	    max
    Error rate:     	9.16%	5.77%	8.51%	0.57%	71.89%
    Insertion rate: 	3.90%	2.80%	3.46%	0.00%	13.32%
    Deletion rate:  	0.97%	1.46%	0.81%	0.00%	60.52%
    Mismatch rate:  	4.29%	2.67%	4.12%	0.00%	13.23%
    Match rate:     	91.80%	5.38%	92.41%	75.06%	100.00%
    Read length:    	2310.96	1518.89	2051.50	97.00	18737.00
    Difference ratio:	48:42:10 (mismatch:insertion:deletion)

The script errorrates.py calculates a detailed error statistics. For our benchmark of RNAseq mapping tools, we have used only mean value for math rate (in bold).

### Runnning Process_pbsim_data.py
Evaluation of simulated data is performed using Process_pbsim_data.py script. It requires files generated by PBSIM during simulation proces.

    python /home/kkrizanovic/src/RNAseqEval/Process_pbsim_data.py process ./simulation_X5/ sample.sam dmelanogaster_chr4.gtf

### Running RNAseqEval.py
Evaluation of real data is performed using RNAseqEval.py script. It requires gene annotations for comparison to mappings.

    python /home/kkrizanovic/src/RNAseqEval/RNAseqEval.py eval-mapping dmelanogaster_chr4_genome.fa sample.sam -a dmelanogaster_chr4.gtf

